### ISSUE TITLE
Style guide not codified 

### ISSUE THREAD

**Issue by renemilk** [wwu-numerik/dune-stuff#13]

**Comment by renemilk** [comment]
Also what's with not having a config for and heavily using clang-format again? See a77a54b41dfba0bcf5a4a0d78d73b76d21fb1b49

**Comment by ftalbrecht** [comment]
At the time there were several things it could not do, that were important to me. But I guess its much better than having nothing. Please do not push a complete reformatting of dune-stuff yet, though. I would like to take a look first.

**Comment by renemilk** [comment]
I've applied it in 228279d1b43bb88204bf1b4ee57d6fcdcbf0d8bf so you can have a look w/o doing anything locally. 
I'm pretty happy with the result. Only minor downside to me is macro indentation getting flattened.
I haven't looked at the clang-format specs in quite a while though, it's quite possibe they've added more options/finer control.

**Comment by ftalbrecht** [comment]
Nice. I would like to take the time and go through the specs of clang-format, probably in August if that's ok (but definitely before the release). But I guess such an automated option will be for the best (as long as some of the definitions are still open for debate...)

**Comment by renemilk** [comment]
Please do. I agree we should apply a whole-module format before the release.  I don't want this release to fade into the background again. So I'd  say we should aim to not trail the core module release by more than a week. 
I also agree that realistically a compromise in clang-format setup and its (semi-)automated application is the only  way we're going to keep a consistent style everywhere.

**Comment by ftalbrecht** [comment]
I will probably have some time next week and will look into this then!
Its even in my calendar now! :)

**Comment by renemilk** [comment]
Core modules released 2.4 8 days ago now. I want to get on with the release. Do you have concrete wishes for changes to the clang-format spec? We can target clang-format 3.7 now btw.
Also anybody got ideas for similarly treating our cmake source files? I've tried https://github.com/rpavlik/cmake-multitool/blob/master/cmake-bulk-decrufter.py but its output is worse than the input more often than not.

**Comment by ftalbrecht** [comment]
There were two nice additions that clang-format 3.7 supports by now, I added them to the `.clang-format` (and forgot to reference this issue in the commit). I am still not 100% happy with the result.

Imho we can just give it a try, given that
* locally, clang-format may be disabled without discussion (if reasonable)
* we allow for going back to the old chaos or a different system in a few month, if unhappy after reviewing/discussing the situation.

I have no idea at all regarding the cmake files, though (but I do not consider this a show stopper before the release).

**Comment by ftalbrecht** [comment]
On a different note: https://github.com/isocpp/CppCoreGuidelines

I did not read through all of them but we might want to just mention them in the `Contributing.md`, along with other references...

**Comment by ftalbrecht** [comment]
Dit you already think about how to implement this?

* I am not sure how easy it will be to merge old branches which are not formatted (e.g., existing in user repos or forks). Do you think this will be unproblematic after first applying the formatting in the old branches as well?
* Should we apply the formatting to all branches we work on currently?
* Should/could we have some kind of automatism which applies the formatting?

**Comment by renemilk** [comment]
- Re your first comment: 297c5ad7f9bcd559ac2a6748bb8dbc79008180e8 only has one change though, right?
- Agreed on locally disabling format. It must stay an exception though imo, and not become a rule. I also think "locally" should never mean an entire file.
- As for going back, what do you mean by that exactly? We can hardly revert the formatting and then re-apply months of changes on top at some point.
- I'll make a separate issue wrt cmake cleanup, I agree that it should not block this issue nor the 2.4 release. It just started bugging me.
- I wasn't aware those guidelines had already been expanded, I viewed them a couple weeks ago and there was hardly anything there. I agree that they're prolly a good candidate for inclusion in the contribution guideline.
- Merging branches back into master might be problematic indeed. Are there feature branches still unmerged atm though? I'm only aware of testtools-refactor, which is tiny wrt actual code change and wouldn't mind.
- If there are ongoing branches it's probably a better idea to merge master after the formatting was applied.
- Yes we *could*. I do have a pre-commit hook somewhere that basically prevents committing if there's a non-empty diff between the actual commit-chunk and clang-format applied to it.

**Comment by ftalbrecht** [comment]
* Yes, forgot the second one, should be in now.
* _Going back_ has to do with how we enforce this policy. I do not mean to alter the code, just that we might stop enforcing this if it is too much trouble for everyone.
* I have the `RS2016` branch, which does differ from master. But I do not think that one should be merged before the release, I guess I will just have to suffer through the merging. It could be that others still have feature branches open, e.g., @tobiasleibner. But again, I am not sure if its good to merge them this soon before the release...
* Regarding enforcing this policy: I think this will be tricky. The behaviour of `clang-format` really depends on the version and I do not think it is a good idea to enforce everyone to have the same clang version installed that we have...

**Comment by tobiasleibner** [comment]
Yes, I do have some features that are not merged yet, mainly functions like ExpressionCheckerboard, a time-dependent Expression function, functions that depend on two variables x and u and a Composition function. But I do not think they should be merged in the current form. Maybe we would like to have a more general interface for multi-variable or parameter-dependent functions replacing the time-dependent and two-variable functions. On the other hand, they are mostly independent from the other stuff and should not break anything else, so we could merge them into the playground or something such that they already get formatted.

**Comment by renemilk** [comment]
I'm pro merging those into playground and massaging the code afterwards, but otoh additional code would be less of potential hassle than modifications to existing. 

I've also had a final (for me) run at the format spec in c39d401.

**Comment by ftalbrecht** [comment]
I was under the impression we are dropping the playground for the release. So what you could do is make a pull request where these go into playground, we then apply the formatting and afterwards remove the playground for the release.
However, I do remember that we did not want to have these functions in their current form since we were aiming for a different approach to handling time dependency...

**Comment by tobiasleibner** [comment]
So the problem with merging old branches is that the format changes would cause a lot of merge conflicts in existing code but new files/new classes would not have that problem, right? I do not think it is worth the effort then with first adding it to the playground and removing it afterwards.

**Comment by ftalbrecht** [comment]
Agreed.

**Comment by renemilk** [comment]
Right. So let's deal with those afterwards and separately. 

Anything else we should discuss wrt the format stuff before I apply it pre-release prep? Aside from finalizing the format spec itself I mean (in #66)

**Comment by renemilk** [comment]
Done 554fc61975c4172cd13af1001b3121b0f48715a0
